opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -541,24 +542,73 @@ namespace Opm
     template <typename TypeTag>
     void
     MultisegmentWell<TypeTag>::
-    computeWellPotentials(const Simulator& /* ebosSimulator */,
-                          const WellState& /* well_state */,
+    computeWellPotentials(const Simulator& ebosSimulator,
+                          const std::vector<Scalar>& B_avg,
+                          const WellState& well_state,
                           std::vector<double>& well_potentials,
                           Opm::DeferredLogger& deferred_logger)
     {
-        const std::string msg = std::string("Well potential calculation is not supported for multisegment wells \n")
-                + "A well potential of zero is returned for output purposes. \n"
-                + "If you need well potential to set the guide rate for group controled wells \n"
-                + "you will have to change the " + name() + " well to a standard well \n";
-
-        deferred_logger.warning("WELL_POTENTIAL_NOT_IMPLEMENTED_FOR_MULTISEG_WELLS", msg);
-
         const int np = number_of_phases_;
         well_potentials.resize(np, 0.0);
 
+        updatePrimaryVariables(well_state, deferred_logger);
+
+        // initialize the primary variables in Evaluation, which is used in computePerfRate for computeWellPotentials
+        // TODO: for computeWellPotentials, no derivative is required actually
+        initPrimaryVariablesEvaluation();
+
+        // get the bhp value based on the bhp constraints
+        const double bhp = Base::mostStrictBhpFromBhpLimits(deferred_logger);
+
+        // does the well have a THP related constraint?
+        if ( !Base::wellHasTHPConstraints() ) {
+            assert(std::abs(bhp) != std::numeric_limits<double>::max());
+
+            computeWellRatesWithBhp(ebosSimulator, B_avg, bhp, /*iterate=*/ true, well_potentials, deferred_logger);
+        } else {
+
+            const std::string msg = std::string("Well potential calculation is not supported for thp controlled multisegment wells \n")
+                    + "A well potential of zero is returned for output purposes. \n"
+                    + "If you need well potential computed from thp to set the guide rate for group controled wells \n"
+                    + "you will have to change the " + name() + " well to a standard well \n";
+
+            deferred_logger.warning("WELL_POTENTIAL_FOR_THP_NOT_IMPLEMENTED_FOR_MULTISEG_WELLS", msg);
+            return;
+        }
+
     }
 
 
+    template<typename TypeTag>
+    void
+    MultisegmentWell<TypeTag>::
+    computeWellRatesWithBhp(const Simulator& ebosSimulator,
+                            const std::vector<Scalar>& B_avg,
+                            const double& bhp,
+                            const bool iterate,
+                            std::vector<double>& well_flux,
+                            Opm::DeferredLogger& deferred_logger)
+    {
+        // store a copy of the well state, we don't want to update the real well state
```

Comment:
Suggestion on how to handle this function is similar to the previous comment with StandardWell. The well_controls_ is not reset here, it is reset in `BlackoilWellModel`. It might not be correct since a THP limit can be there at the same time. 

Eventually, if our code can just set the BHP and THP together, then we just solve it, the resulting well rates will be the well potentials, that will be great and that should be our goal, while I do not think we can do that in this PR, at least not for MSW.  We still have to calculate under BHP and THP separately and pick the one gives the smaller well potential value.